extensions/vscode/src/entrypointTracker.ts

```diff
@@ -0,0 +1,214 @@
+// Copyright (C) 2023 by Posit Software, PBC.
+
+import {
+  Disposable,
+  TextDocument,
+  TextEditor,
+  commands,
+  window,
+  workspace,
+} from "vscode";
+import { Utils as uriUtils } from "vscode-uri";
+
+import { useApi } from "src/api";
+import { Contexts } from "src/constants";
+import { getPythonInterpreterPath } from "src/utils/config";
+import { isActiveDocument, relativeDir } from "src/utils/files";
+import { hasKnownContentType } from "src/utils/inspect";
+import { getSummaryStringFromError } from "src/utils/errors";
+
+/**
+ * Determines if a text document is an entrypoint file.
+ *
+ * @param document The text document to inspect
+ * @returns If the text document is an entrypoint
+ */
+async function isDocumentEntrypoint(document: TextDocument): Promise<boolean> {
+  const dir = relativeDir(document.uri);
+  // If the file is outside the workspace, it cannot be an entrypoint
+  if (dir === undefined) {
+    return false;
+  }
+
+  try {
+    const api = await useApi();
+    const python = await getPythonInterpreterPath();
+
+    const response = await api.configurations.inspect(python, {
+      dir: dir,
+      entrypoint: uriUtils.basename(document.uri),
+    });
+
+    return hasKnownContentType(response.data);
+  } catch (error: unknown) {
+    const summary = getSummaryStringFromError(
+      "entrypointTracker::isDocumentEntrypoint",
+      error,
+    );
+    window.showInformationMessage(summary);
+    return false;
+  }
+}
+
+/**
+ * Tracks whether a document is an entrypoint file and sets extension context.
+ */
+export class TrackedEntrypointDocument {
+  readonly document: TextDocument;
+  private isEntrypoint: boolean;
+
+  private requiresUpdate: boolean = false;
+
+  private constructor(document: TextDocument, isEntrypoint: boolean) {
+    this.document = document;
+    this.isEntrypoint = isEntrypoint;
+  }
+
+  static async create(document: TextDocument) {
+    const isEntrypoint = await isDocumentEntrypoint(document);
+    return new TrackedEntrypointDocument(document, isEntrypoint);
+  }
+
+  /**
+   * Sets the file entrypoint context with this as the active file.
+   * @param options Options for the activation
+   * @param options.forceUpdate Whether to force the entrypoint to update
+   */
+  async activate(options?: { forceUpdate?: boolean }) {
```

Comment:
The API calls only happen when we create a `TrackedEntrypointDocument` or when it is activated and needs an updated.

This occurs when the document is saved, or the document is re-opened.

`TrackedEntrypointDocument`s don't update if the file is open in VSCode, another file becomes active, then the original file is active again. That is because nothing has changed since the first API call.